src/main/java/org/springframework/social/partnercenter/security/AzureADMultiTenantOAuthTemplate.java

```diff
@@ -0,0 +1,251 @@
+import org.springframework.social.partnercenter.http.logging.LogLevel;
+import org.springframework.social.partnercenter.http.logging.LoggingRequestInterceptor;
+import org.springframework.social.partnercenter.serialization.Json;
+import org.springframework.social.partnercenter.serialization.JsonSerializationException;
+import org.springframework.social.support.ClientHttpRequestFactorySelector;
+import org.springframework.social.support.FormMapHttpMessageConverter;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.HttpStatusCodeException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+
+public class AzureADMultiTenantOAuthTemplate implements AzureADMultiTenantOAuthOperations {
+	private final String clientId;
+	private final String clientSecret;
+	private RestTemplate restTemplate;
+	private final UriProvider uriProvider;
+
+	/**
+	 * Constructs an OAuth2Template for a given set of client credentials.
+	 * @param clientId the multi tenant application ID
+	 * @param clientSecret the multi tenant application secret
+	 * @param uriProvider builds auth urls
+	 */
+	public AzureADMultiTenantOAuthTemplate(String clientId, String clientSecret, UriProvider uriProvider) {
+		notNull(clientId, "The clientId property cannot be null");
+		notNull(clientSecret, "The clientSecret property cannot be null");
+		this.uriProvider = ofNullable(uriProvider).orElse(DEFAULT_URL_PROVIDER);
+		this.clientId = clientId;
+		this.clientSecret = clientSecret;
+	}
+
+	@Override
+	public String buildAuthorizeUrl(String clientId, String redirectUri, String state) {
+		return uriProvider.buildAuthorizeUrl(clientId, redirectUri, state);
+	}
+
+	@Override
+	public String buildAuthorizeUrl(String clientId, OAuth2Parameters additionalParams) {
+		return uriProvider.buildAuthorizeUrl(clientId, additionalParams);
+	}
+
+	@Override
+	public String exchangeForRefresh(String authorizationCode, String partnerTenantId, String redirectUri, MultiValueMap<String, String> additionalParameters) {
+		MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+		params.set("client_id", clientId);
+		params.set("client_secret", clientSecret);
+		params.set("code", authorizationCode);
+		params.set("redirect_uri", redirectUri);
+		params.set("grant_type", "authorization_code");
+
+		ofNullable(additionalParameters).ifPresent(additionalParameterMap ->
+				additionalParameterMap.entrySet().stream()
+						.filter(entry -> !ImmutableList.of("client_id", "redirect_uri", "code", "client_secret", "grant_type").contains(entry.getKey()))
+						.forEach(entry -> params.put(entry.getKey(), entry.getValue())));
+
+		UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder
+				.fromUriString("https://login.microsoftonline.com")
```

Comment:
we don't still have the authority in the security region? This won't work for DE

Reply:
Right but the authority is actually wrong for US at least in this flow. The current authority is `USA("https://login.windows.net", "https://graph.windows.net", "https://api.partnercenter.microsoft.com"),` [SecurityRegion.java](https://github.com/SpringSocialMicrosoft/spring-social-partnercenter/blob/master/src/main/java/org/springframework/social/partnercenter/api/uri/SecurityRegion.java#L7), but the authority we need to use in this flow is `https://login.microsoftonline.com`. I was unsure about changing that authority since old flows use them and the refresh access token flow for a partner center connection also does. 

Maybe the right solution here could be to have a different authority for oauth flows? Or would simply replacing it still work?

Reply:
After testing it, both urls redirect to `https://login.microsoftonline.com`. maybe I will change the authority to `https://login.microsoftonline.com`